python/instrumentation/openinference-instrumentation-llama-index/tests/openinference/instrumentation/llama_index/test_handler.py

```diff
@@ -0,0 +1,521 @@
+import asyncio
+import json
+import logging
+import random
+from collections import defaultdict
+from concurrent.futures import ThreadPoolExecutor
+from contextlib import suppress
+from itertools import count
+from typing import (
+    Any,
+    AsyncIterator,
+    DefaultDict,
+    Dict,
+    Generator,
+    Iterable,
+    Iterator,
+    List,
+    Optional,
+    Tuple,
+    cast,
+)
+
+import openai
+import pytest
+from httpx import AsyncByteStream, Response, SyncByteStream
+from llama_index.core import Document, ListIndex, Settings
+from llama_index.core.base.response.schema import StreamingResponse
+from llama_index.core.callbacks import CallbackManager
+from llama_index.core.schema import TextNode
+from llama_index.llms.openai import OpenAI  # type: ignore
+from openinference.instrumentation import using_attributes
+from openinference.instrumentation.llama_index import LlamaIndexInstrumentor, _legacy_llama_index
+from openinference.semconv.trace import (
+    DocumentAttributes,
+    EmbeddingAttributes,
+    MessageAttributes,
+    OpenInferenceMimeTypeValues,
+    OpenInferenceSpanKindValues,
+    SpanAttributes,
+    ToolCallAttributes,
+)
+from opentelemetry import trace as trace_api
+from opentelemetry.sdk import trace as trace_sdk
+from opentelemetry.sdk.trace import ReadableSpan
+from opentelemetry.sdk.trace.export import SimpleSpanProcessor
+from opentelemetry.sdk.trace.export.in_memory_span_exporter import InMemorySpanExporter
+from respx import MockRouter
+
+for name, logger in logging.root.manager.loggerDict.items():
+    if name.startswith("openinference.") and isinstance(logger, logging.Logger):
+        logger.setLevel(logging.DEBUG)
+        logger.handlers.clear()
+        logger.addHandler(logging.StreamHandler())
+
+
+@pytest.mark.skipif(
+    _legacy_llama_index(),
+    reason="legacy callback system",
+)
+@pytest.mark.parametrize(
+    "is_stream,is_async",
+    [
+        (False, False),
+        (False, True),
+        (True, False),
+        # FIXME: stream + async is not supported by LlamaIndex as of v0.9.33
+        # (True, True),
+    ],
+)
+@pytest.mark.parametrize(
+    "status_code", [200]
+)  # 400 has been removed because retries have been added and can't be easily modified
+@pytest.mark.parametrize("use_context_attributes", [False, True])
```

Comment:
I think we're overusing parameterization here. Parameterization is definitely useful for many test cases, but in this case, it introduces complex branching logic that makes the test case harder to understand. I like the way this comment puts it: [DRY is for code, DAMP is for tests](https://stackoverflow.com/questions/6453235/what-does-damp-not-dry-mean-when-talking-about-unit-tests/11837973#11837973).

Reply:
This is a bit of trade off because the code that is doing the checking is fairly massive, and if we abstract it out as a helper function, `pytest` doesn't seem to do a very good job of pin-pointing the exact assert statement that is failing.

Reply:
Definitely a challenge. It looks like there's a workaround [here](https://stackoverflow.com/questions/41522767/pytest-assert-introspection-in-helper-function). If we have multiple test cases, we also don't necessarily need to check every attribute for every test case, only the ones that are important to the particular test case.

This is non-blocking.